conversion/conversion.go

```diff
@@ -0,0 +1,789 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Package conversion handles initial setup for the command line tool
+// and web APIs.
+
+package conversion
+
+import (
+	"bufio"
+	"context"
+	"crypto/rand"
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"strings"
+	"sync/atomic"
+	"syscall"
+	"time"
+
+	sp "cloud.google.com/go/spanner"
+	database "cloud.google.com/go/spanner/admin/database/apiv1"
+	instance "cloud.google.com/go/spanner/admin/instance/apiv1"
+	"github.com/aws/aws-sdk-go/aws/session"
+	dydb "github.com/aws/aws-sdk-go/service/dynamodb"
+	_ "github.com/go-sql-driver/mysql"
+	_ "github.com/lib/pq"
+	"golang.org/x/crypto/ssh/terminal"
+	"google.golang.org/api/iterator"
+	adminpb "google.golang.org/genproto/googleapis/spanner/admin/database/v1"
+	instancepb "google.golang.org/genproto/googleapis/spanner/admin/instance/v1"
+
+	"github.com/cloudspannerecosystem/harbourbridge/dynamodb"
+	"github.com/cloudspannerecosystem/harbourbridge/internal"
+	"github.com/cloudspannerecosystem/harbourbridge/mysql"
+	"github.com/cloudspannerecosystem/harbourbridge/postgres"
+	"github.com/cloudspannerecosystem/harbourbridge/spanner"
+	"github.com/cloudspannerecosystem/harbourbridge/spanner/ddl"
+)
+
+const (
+	// PGDUMP is the driver name for pg_dump.
+	PGDUMP string = "pg_dump"
+	// POSTGRES is the driver name for PostgreSQL.
+	POSTGRES string = "postgres"
+	// MYSQLDUMP is the driver name for mysqldump.
+	MYSQLDUMP string = "mysqldump"
+	// MYSQL is the driver name for MySQL.
+	MYSQL string = "mysql"
+	// DYNAMODB is the driver name for AWS DynamoDB.
+	// This is an experimental driver; implementation in progress.
+	DYNAMODB string = "dynamodb"
+)
+
+func SchemaConv(driver string, ioHelper *IOStreams, schemaSampleSize int64) (*internal.Conv, error) {
+	switch driver {
+	case POSTGRES, MYSQL:
+		return schemaFromSQL(driver)
+	case PGDUMP, MYSQLDUMP:
+		return schemaFromDump(driver, ioHelper)
+	case DYNAMODB:
+		return schemaFromDynamoDB(schemaSampleSize)
+	default:
+		return nil, fmt.Errorf("schema conversion for driver %s not supported", driver)
+	}
+}
+
+// Check if schema contains interleaved tables.
+func checkInterleaved(conv *internal.Conv) bool {
+	for _, table := range conv.SpSchema {
+		if table.Parent != "" {
+			return true
+		}
+	}
+	return false
+}
+
+func DataConv(driver string, ioHelper *IOStreams, client *sp.Client, conv *internal.Conv, dataOnly bool) (*spanner.BatchWriter, error) {
+	config := spanner.BatchWriterConfig{
+		BytesLimit: 100 * 1000 * 1000,
+		WriteLimit: 40,
+		RetryLimit: 1000,
+		Verbose:    internal.Verbose(),
+	}
+	switch driver {
+	case POSTGRES, MYSQL:
+		return dataFromSQL(driver, config, client, conv)
+	case PGDUMP, MYSQLDUMP:
+		if checkInterleaved(conv) {
+			return nil, fmt.Errorf("Schema contains interleaved tables")
```

Comment:
I think we need a bit more explanation here for users, including suggested workaround.

Either expand the error message, or precede it with something like:

fmt.Fprintf(ioHelpter.out, "HarbourBridge does not currently support data conversion from dump files\nif the schema contains interleaved tables. Suggest using direct access to source database\ni.e. using drivers postgres and mysql.")

Reply:
updated error message